ensure cart totals update when items are removed or quantity changed

haszari opened the pull request:
Fixes #1830

~~This PR modifies the `DELETE /cart/items/:key` endpoint so that it always returns a complete cart object (previously returns no content).~~

This PR moves the `DELETE /cart/items/:key` endpoint to cart controller as `POST /cart/remove-item/?key=` so that it always returns a complete cart object (previously returns no content). Also moves the API for updating item quantity to `POST /cart/update-item`. The client actions, API docs and PHPunit tests have been updated alongside this change. 

The [idea here is that any endpoint that modifies the cart returns full cart](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/1830#issuecomment-592469114), as other data may change (e.g. add a coupon => totals change, remove an item => totals change, shipping may change etc). 

Note that the only two APIs that have been moved are for deleting and removing cart items. I've moved these specifically because we need them to return full cart for the cart block to update totals (etc). There are other routes which could be moved, but we can discuss / do that as follow up - see related issue #1849.


### How to test the changes in this Pull Request:

1. Add a cart block to a page and publish.
2. Add some stuff to cart.
3. View cart page and remove items – totals and other cart details should be correct.
4. View cart page and change item quantities - totals and other details should update as items change.

mikejolley commented:
@nerrad @haszari I'm not super keen on this approach because this **does not return the correct schema.** `/cart/items` has a different schema to `/cart`, so if anything were to be returned here it would be a cart item -- but we don't want that.

The way I see it we have 2 options:

1. Similar to some of the other endpoints, move cart manipulation endpoints under `/cart` so we can return a cart as the response.

2. Just chain the API queries. This issue could be fixed by implementing a second request to get a fresh cart - https://github.com/woocommerce/woocommerce-gutenberg-products-block/compare/fix/1830-remove-titles-on-remove-alt

nerrad commented:
> I'm not super keen on this approach because this does not return the correct schema. /cart/items has a different schema to /cart, so if anything were to be returned here it would be a cart item -- but we don't want that.

I agree with Mike here.

Regarding the two options I think I prefer option one to keep the request count down. It also follows the convention we're starting to establish where all cart manipulations affecting totals and needing a fresh cart instance happen via the `cart` endpoint.

The only downside is that we're straying from REST principles and `/cart` is becoming more of an RPC type endpoint (which is okay, just as long we acknowledge that).

haszari commented:
@nerrad @mikejolley That makes sense to me - I'll rework this so the cart-modifying endpoints (e.g. change quantity, delete item) are on `/cart` controller and return full cart.

haszari commented:
@mikejolley @nerrad I've updated this so the remove item endpoint is on the Cart controller. Ready for another review.

I'm thinking I should use this same approach for cart item quantity endpoint - move this to the Cart controller, return full cart, and then update the action so totals update correctly (and other dependent stuff). Let me know what you think - i.e. could do as separate PRs.

haszari commented:
> Update: I can't get the PHP api test working, I don't understand why it's failing. Any ideas?

Update update! My test was doing two things:

1. Remove a valid item, assert that worked correctly.
2. Attempt to remove same item again, assert that this fails with 404.

I now have a separate test similar to (2), which attempts to remove an invalid item and asserts 404.

So there seems to be an issue with running both those tests in one test, so I'll split them into two different tests, which seems cleaner (though it's now testing much simpler scenario).

haszari commented:
Still having trouble with the tests, and for some reason Travis isn't showing me logs. For now I've removed the troublesome test (`test_remove_cart_item`)

<img width="1024" alt="Screen Shot 2020-03-03 at 4 35 33 PM" src="https://user-images.githubusercontent.com/4167300/75740605-041c3f80-5d6d-11ea-8b76-3050f8430c2d.png">

haszari commented:
I'm going to press on with this - take the same approach with the quantity endpoint. Seems like we're all in agreement this is the right approach (for now at least). cc @mikejolley @nerrad

haszari commented on src/RestApi/StoreApi/README.md:
````diff
@@ -722,66 +753,6 @@ Example response:
-| `quantity` | integer |   Yes    | Quantity of this item in the cart. |
-
-```http
-curl --request PUT https://example-store.com/wp-json/wc/store/cart/items/3ef815416f775098fe977004015c6193&quantity=10
````
Note: I've removed these `curl` examples when moving docs because they don't work (e.g. endpoint path is a param in WP rest API). @mikejolley what do you think about removing the curl examples throughout? Let me know if they work for you.

mikejolley replied:
Can we not just fix the examples if you see a problem? Otherwise we cannot document a request and response.

haszari replied:
Sure - I'll get these working and post and update. I definitely think it's a nice idea to have pasteable examples.

Just realised why these aren't working for me – I had plain permalinks option set. I'll add a reminder about that (gotcha) to the doc.

haszari commented on src/RestApi/StoreApi/Controllers/Cart.php:
```diff
@@ -103,6 +103,44 @@ public function register_routes() {
+			'/' . $this->rest_base . '/update-item',
+			[
+				[
+					'methods'  => RestServer::EDITABLE,
```
This route supports `EDITABLE` (i.e. `POST | PUT | PATCH`), consistent with the previous API from `cart/items`. @mikejolley I'm thinking we could use `EDITABLE` for all these cart modifier routes - thoughts? Alternatively could make them all `POST` only for consistency.

haszari replied:
Update - this is now `POST` for consistency with other `/cart` routes.

haszari commented:
This PR is now ready for review 🎉FYI @mikejolley @nerrad 

- Both cart item endpoints used for removing/deleting are now on Cart controller and return full cart.
- Have updated the API tests (now passing) and documentation.
- Totals and other dependent cart data should now update when removing or changing quantities.

haszari commented:
This PR is ready for review again - I've addressed @mikejolley feedback:

- Reinstated previous `CartItems` endpoints for deleting/updating item quantity. Cart isn't using these right now FYI. Related issue: #1849
- Added curl examples to docs for new endpoints, and tip about not using plain permalinks setting.

There are two Travis jobs failing, I can't figure out if they are due to my changes. 

1. JS/CSS lint & bundle size. There’s a permission issue with `cross-env`, which makes me think it’s an issue in travis setup/environment. `npm run build:ci` works locally.
2. E2E tests. These are failing on checkout snapshot test - when I run locally it passes. Though it makes sense to me that we'd need to regenerate snapshots, those tests just got merged and the cart markup is still in development.

senadir commented:
I rebased this to master so tests can pass

haszari commented:
> I restarted the build/lint job in travis and it passed. When I see weird errors like that on travis I usually first try a restart of the job before doing any other debugging. About 90% of the time it's just a problem with that particular travis build and restarting works fine.

I reran the job quite a few times yesterday! 🎉 that it's working now. Merging :)
